lib/std/Thread/RwLock.zig

```diff
@@ -246,4 +247,132 @@ pub const DefaultRwLock = struct {
+    try testing.expect(!rwl.tryLock());
+    try testing.expect(!rwl.tryLockShared());
+    rwl.unlock();
+
+    try testing.expect(rwl.tryLock());
+    try testing.expect(!rwl.tryLock());
+    try testing.expect(!rwl.tryLockShared());
+    rwl.unlock();
+
+    rwl.lockShared();
+    try testing.expect(!rwl.tryLock());
+    try testing.expect(rwl.tryLockShared());
+    rwl.unlockShared();
+    rwl.unlockShared();
+
+    try testing.expect(rwl.tryLockShared());
+    try testing.expect(!rwl.tryLock());
+    try testing.expect(rwl.tryLockShared());
+    rwl.unlockShared();
+    rwl.unlockShared();
+
+    rwl.lock();
+    rwl.unlock();
+}
+
+test "RwLock - concurrent access" {
+    if (builtin.single_threaded)
+        return;
+
+    const num_writers: usize = 2;
+    const num_readers: usize = 4;
+    const num_writes: usize = 10000;
+
+    const Runner = struct {
+        const Self = @This();
+
+        rwl: RwLock = .{},
+        writes: usize = 0,
+        reads: std.atomic.Atomic(usize) = std.atomic.Atomic(usize).init(0),
+
+        term1: usize = 0,
+        term2: usize = 0,
+        term_sum: usize = 0,
+
+        fn reader(self: *Self) !void {
+            while (true) {
+                self.rwl.lockShared();
+                defer self.rwl.unlockShared();
+                if (self.writes >= num_writes) break;
+
+                try self.check();
+
+                _ = self.reads.fetchAdd(1, .AcqRel);
+            }
+        }
+
+        fn writer(self: *Self, thread_idx: usize) !void {
+            var prng = std.rand.DefaultPrng.init(thread_idx);
+            var rnd = prng.random();
+
+            while (true) {
+                self.rwl.lock();
+                defer self.rwl.unlock();
+                if (self.writes >= num_writes) break;
+
+                try self.check();
+
+                const term1 = rnd.int(usize);
+                self.term1 = term1;
+                try std.Thread.yield();
+
+                const term2 = rnd.int(usize);
+                self.term2 = term2;
+                try std.Thread.yield();
+
+                self.term_sum = term1 +% term2;
+                self.writes += 1;
+            }
+        }
+
+        fn check(self: *const Self) !void {
+            const term_sum = self.term_sum;
+            try std.Thread.yield();
+
+            const term2 = self.term2;
+            try std.Thread.yield();
+
+            const term1 = self.term1;
+            try testing.expectEqual(term_sum, term1 +% term2);
+        }
+    };
+
+    var runner = Runner{};
+    var threads: [num_writers + num_readers]std.Thread = undefined;
+
+    for (threads[0..num_writers]) |*t, i| t.* = try std.Thread.spawn(.{}, Runner.writer, .{ &runner, i });
+    for (threads[num_writers..]) |*t| t.* = try std.Thread.spawn(.{}, Runner.reader, .{&runner});
+
+    for (threads) |t| t.join();
+
+    try testing.expectEqual(num_writes, runner.writes);
+    try testing.expect(runner.reads.load(.SeqCst) >= 10 * num_readers);
```

Comment:
Why should each reader have done at least 10 reads specifically?

Also this could be Monotonic/Relaxed as well. The increments are bound to be observable due to the thread joins (they're implicitly Acquire on the thread's Release after completion).

Reply:
It is just checking that a few reads were granted as opposed to none, which would indicate a bug.

Reply:
It's problematic as all writers could run to completion (yield could noop in the scheduler), then the readers could run once -> observe writing is finished -> exit before incrementing making `reads=0` after everything is joined.

Reply:
Good point, I will change all access to `runner.reads` to Unordered after the current tests finish.

Reply:
Unordered isn't allowed for rmw so I used Acquire.

Reply:
I will remove the check at the end for non-zero reads to avoid CI failures under extreme conditions.

Reply:
Unordered doesn't guarantee modification order (read: updates which observe the "latest" value for lack of better words), only that the access is atomic. This means they dont work with CAS or any other rmw. You could use Monotonic instead as there's no happens-before needed for Acquire to be required.

Reply:
Thanks, I'll use Monotonic.